### PR TITLE
cEP-0025.md: Integrate pyflakes AST

### PR DESCRIPTION
**Pull request by ankitxjoshi**
Closes https://github.com/coala/cEPs/issues/143

### CODE REVIEW

**Comment by ankitxjoshi** [comment]
pseudocode for plugin implementation yet to be added.

File: cEP-0025.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                                               result.deadScopes),
+                    'ModuleScope': get_scope(ModuleScope,
+                                             result.deadScopes),
+                    'GeneratorScope': get_scope(GeneratorScope,
+                                                result.deadScopes),
+                    'DoctestScope': get_scope(DoctestScope,
+                                              result.deadScopes)
+        }
+
+        yield HiddenResult(self, content)
```
Comment (jayvdb) [inline cEP-0025.md]: create a HiddenResult subclass, and use OO. a dict is ugly, and brittle.  backwards compatibility and deprecation with a dict is limited, so a dict will mean we have problems when the pyflakes internals change.

Also please see all the issues related to HiddenResults.  Fixing a few of them in your project will make everything much neater.

**Review by jayvdb — CHANGES_REQUESTED** [review]

**Comment by ankitxjoshi** [comment]
I believe that the flake8 wrapper (another part of the project) should be included in a different cEP.

File: cEP-0025.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+| Type     | Feature                                         |
+
+## Abstract
+
+This document describes how to create a metabear to integrate
+pyflakes-enhanced-AST into coala.
+
+## Introduction
+
+The current approach of plugin development for flake8 does not utilises
```
Comment (jayvdb) [inline cEP-0025.md]: This jumps straight to `flake8`, without helping the reader understand how flake8 is related to this coala project.

You might want to start with something like flake8 is the most commonly used linter framework in the python world. (see https://pythonwheels.com/ for a list of most downloaded pypi packages, and there are other pypi stats pages which might be more appropriate to use.) flake8 has two builtin enabled linters, pyflakes and pycodestyle.
Additional linters are supported using the flake8 plugin system, and may be Python AST based or built using pycodestyle's internal logical-line-based checker API.  There has not been an option to use the pyflakes internal AST based checker API to create custom linters.

(might be worth mentioning https://github.com/flintwork/flint for history sake..)

Then you can explain why using the pyflakes internal AST based checker API to create custom linters is a fantastic idea .. ;-)

To cover the pycodestyle's internal logical-line-based checker API, the best example is https://github.com/openstack-dev/hacking , which has lots of little rules.

It would also be useful to investigate the https://github.com/PyCQA/bandit plugin system (recently added to PyCQA), as that is AST based iirc.

It wouldnt hurt to also mention pylint, which has plugin support and the best Python ast IMO, [astroid](https://github.com/PyCQA/astroid).  And it would definitely be a good idea to document why we are using pyflakes instead of astroid.

One of the reasons is along the lines of what you've already said - pyflakes provides a lot of re-usable analysis, making plugins easy, while astroid is more detailed and powerful, it is much more complicated.  And there is a much larger ecosystem of flake8 plugins.

A project objective not relevant for this cEP is that we may be able to get flake8 to add a new 'pyflakes ast' plugin type, which has the same API as our own, or a similar enough API that the wrapper is very thin.

File: cEP-0025.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+the implementation of the new logic that his/her plugin provides
+and not about the fidelity of the basic node handlers.
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
+2. The repository will contain a new package `pyflakes-plugins` which
```
Comment (jayvdb) [inline cEP-0025.md]: If this package contains primarily *bears*, it needs to include either *bears* or *coala* in the name.

But it might be worth creating a non-coala API for these linter plugins - an API which is similar to other frameworks like flake8, so they could also run the same linters.

Then the package contains a bunch of generic pyflakes plugins, a coala meta bear (the pyflakes ast provider), and a coala bear meta wrapper (which creates a wrapper bear for each of the plugins).

Or we could have two packages, one for the generic pyflakes plugins, and a different one for the coala glue.

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0025.md (hunk 2 above)
Comment (jayvdb) [reply cEP-0025.md]: Actually the last sentence isnt quite correct.  The [original project](http://projects.coala.io/#/projects?project=integrate_pyflakes-enhanced_ast_into_coala_&lang=en) has a goal of "invoke existing flake8 plugins" (which would be the pycodestyle API and raw python ast plugin types which flake8 supports) which is beyond the scope of this pyflakes cEP

However the project also mentions "This allows the coala and flake8 communities to cooperate on plugin development using a common plugin framework.", so it is within the scope of this project to ensure that the new pyflakes ast plugins created are also usable by flake8.  We might supply a patch to flake8 to support the new pyflakes ast plugin type, or this repo/package might provide a flake8-plugin which invokes all of the new pyflakes-ast plugins in this repo.

coala already has `PyFlakesBear` and `PycodestyleBear`, so we dont need a Flake8 bear for its default plugins.  But there are lovely flake8 existing plugins which we would like to have as separate bears.

See https://github.com/coala/coala-bears/issues/1388 for the start of a rabbit hole about why we dont want to depend on flake8 directly.  And this problem would also occur if we depended on any flake8 plugins as they will give flake8 as a dependency, and then we are stuck with flake8 dictating our versions.  And this problem would occur if this package of flake8-pyflakes-ast-plugins added a dependency on flake8.  How to solve this dependency problem is an implementation detail which can be omitted from this cEP.

File: cEP-0025.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+Here is a prototype for the implementations of PyflakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class PyflakesResult(HiddenResult):
+
+    @enforce_signature
+    def __init__(self, origin, deadScopes):
```
Comment (jayvdb) [inline cEP-0025.md]: enforce signature depends on having type annotations here.

File: cEP-0025.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scope = self.get_scope(ClassScope, deadScopes)
+        self.function_scope = self.get_scope(FunctionScope, deadScopes)
+        self.module_scope = self.get_scope(ModuleScope, deadScopes)
+        self.generator_scope = self.get_scope(GeneratorScope, deadScopes)
+        self.doctest_scope = self.get_scope(DoctestScope, deadScopes)
+
+    def get_scope(self, scope_level, scopes):
```
Comment (jayvdb) [inline cEP-0025.md]: get_scope -> get_scopes ; always use plural to indicate a list is returned.

and then update the instance attributes.

File: cEP-0025.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    def run(self, filename, file,
+            dependency_results=dict()
+            ):
+        for result in dependency_results.get(PyflakesASTBear.name, []):
+            for scope in result.module_scope:
+                for node_name, node in scope.items():
+                    if isinstance(node, FutureImportation):
+                        yield Result.from_values(
+                            origin=self,
+                            message=('Future import %s found' % node_name),
```
Comment (jayvdb) [inline cEP-0025.md]: this example assumes creates an error for any future imports.

pls investigate the new py37 future import, to see if it is desirable.

If yes, the logic needs to be improved.

If not, then this is a `NoFutureImportBear`.

File: cEP-0025.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+The current approach of plugin development for flake8 does not utilises
+the full potential of enhanced AST rather they simply depend on python AST
+and so, a whole lot of rework is required to do the basic traversing
+and collection of important nodes.
+
+Pyflakes provides with a basic API that does the traversing.
+So, if a developer uses enhanced AST he just needs to work on
+the implementation of the new logic that his/her plugin provides
+and not about the fidelity of the basic node handlers.
```
Comment (jayvdb) [inline cEP-0025.md]: and important problem not yet discussed is that pyflakes internals show the state with erroneous objects as-is in the scopes.

That could lead to plugins doing the 'wrong' thing.  Usually the wrong thing has been done in the code, and needs to be fixed in the source, and then coala restarted, so

1. maybe no pyflakes plugins are run unless pyflakes found no errors.
2. erroneous objects are detected (does pyflakes keep enough metadata to do this?), and then either removed (creating more problems?) or flagged (meaning plugin authors need to check a flag..ugh).

File: cEP-0025.md (hunk 6 above)
Comment (ankitxjoshi) [reply cEP-0025.md]: I was planning to make this bear work 2 ways: 
1. To make sure that a particular import is added to every module. For eg: If I want that print_function should be imported into every module, I would make it compulsory and use this to find instances where I have failed to add them.
2. If I don't want them, or don't want some (create a blacklist). The NoFutureImportBear you are discussing.

Comment (jayvdb) [reply cEP-0025.md]: As an example `NoFutureImportBear` is better.
It is a simple bear that coala can use, as there are no py3 future imports which are supported on py34.  Just add a comment which explains why this bear is simple.
An objective of this cEP and project is to show that it will allow simple rules based on the pyflakes ast.
Making it a complicated bear detracts from that goal.

**Comment by ankitxjoshi** [comment]
Need pointers for `node_visitor` and `scope_visitor` that you discussed.

File: cEP-0025.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+from the 2.5 series today, you will not find new checks in the same series
+in a month from now when you install it again.
+
+It even provides with a plugin mechanism. Additional linters are
+supported using the flake8 plugin system, and may be Python AST
+based or built using pycodestyle's internal logical-line-based
+checker API (checkout [Hacking](https://github.com/openstack-dev/hacking)).
+
+There has not been an option to use the pyflakes internal AST based checker
+API to create custom linters. Thus, the full potential of enhanced AST isn't
```
Comment (jayvdb) [inline cEP-0025.md]: `enhanced AST` here vs `pyflakes internal AST` on the previous line.

The reader may not get the leap from `pyflakes internal AST` to `enhanced AST` being the same thing.

File: cEP-0025.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+and provision of lots of re-usable analysis. Asteroid is more detailed and
+powerful, thus making it complicated. Plus, there is a much larger ecosystem
+of flake8 plugins.
+
+## Proposed Change
+
+Here is a brief overview of the architecture:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
```
Comment (jayvdb) [inline cEP-0025.md]: It seems like we have three packages in one repo.

- pyflakes-generic-plugins is easy -- it contains the generic plugins, and other generic stuff
- pyflakes-plugin-bears is pyflakes enhanced AST coala Bears.

If `pyflakes-plugin-bears` contains metabear `PyFlakesASTBear` ...

What is in `coala-pyflakes` ?

File: cEP-0025.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+1. PyFlakesASTBear will be LocalBear returning result of the type
+   HiddenResult since these results are not meant for the users but
+   for the dependent bear. The result structure is described by
+   PyFlakesResult class.
+2. PyFlakesResult consists of a snapshot of the module at class,
+   function, module, generator ,and doctest level. Thus, providing
+   the developer multiple views of a file allowing him to retrieve
+   results as per his demand.
+3. Apart from scope information, the metabear even returns a list of
```
Comment (jayvdb) [inline cEP-0025.md]: even -> also

File: cEP-0025.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+```python
+# Imports not mentioned to maintain brevity
+
+class PyFlakesResult(HiddenResult):
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scope = self.get_scopes(ClassScope, deadScopes)
```
Comment (jayvdb) [inline cEP-0025.md]: all of these, except module_scope, should be plural: `foo_scopes`.

File: cEP-0025.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        self.function_scope = self.get_scopes(FunctionScope, deadScopes)
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)
+        self.generator_scope = self.get_scopes(GeneratorScope, deadScopes)
+        self.doctest_scope = self.get_scopes(DoctestScope, deadScopes)
+        self.pyflakes_messages = pyflakes_messages
+
+    def get_scopes(self, scope_level, scopes):
```
Comment (jayvdb) [inline cEP-0025.md]: scope_level -> scope_type.

many of these scopes can be nested in various ways, such as functions in a doctest.

We'll probably want to add some helper which makes it easier for plugins to skip anything in a doctest, as they are usually more complicated, and need different logic.

And this is why we probably need to support a visitor pattern, as plugins think about scopes from outer to the inner, and want to exclude large branches that dont interest them.  But the visitor pattern can be thought about later after we've built some plugins and find the pain points.

Actually this is something that is worth mentioning in more detail above.  I am pretty sure Python ast doesnt provide doctests, and pretty sure that pycodestyle also doesnt have any reasonable way to create plugins for doctest.  The pyflakes ability to provide doctest is a serious feature for projects like coala who use them extensively.  Currently errors in them are caught by pyflakes, but the style of them is mostly ignored.

Maybe this is another good example for the cEP, as you could easily show a plugin which is probably very difficult in any other framework.

File: cEP-0025.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            type of scope and values being a list of scopes generated
+            from the file.
+        """
+        tree = ast.parse(''.join(file))
+        result = Checker(tree, filename)
+
+        yield PyFlakesResult(self, result.deadScopes, result.messages)
+```
+
+## Demonstrating use of PyFlakesASTBear by creating a plugin
```
Comment (jayvdb) [inline cEP-0025.md]: plugin -> simple bear.

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0025.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+This document describes how to create a metabear to integrate
+pyflakes-enhanced-AST into coala and demonstrates its use by
+creating a simple dependent bear.
+
+## Introduction
+
+Flake8 is the most commonly used linter framework when it comes to static
+code analysis for python. It basically wraps around the tools pyflakes,
+pycodestyle and mccabe scripts because each of those projects
```
Comment (jayvdb) [inline cEP-0025.md]: > because each ...

That because is a bit strange.
it wraps then because they are useful tools with difference interfaces.
flakes v2 wasnt just a pycodestyle wrapper . It also used pycodestyle engine , plugin system, config system.  It was a hack on pycodestyle.  pyflakes doesnt provide error codes. Flake8 does, and that is why ppl used it , and built new plugins.

Comment (jayvdb) [reply cEP-0025.md]: The fact that new checks are added in minor releases isnt significant to your cEP. Explaining it properly is too much unnecessary work.

File: cEP-0025.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+class PyFlakesResult(HiddenResult):
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scopes = self.get_scopes(ClassScope, deadScopes)
+        self.function_scopes = self.get_scopes(FunctionScope, deadScopes)
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)
```
Comment (jayvdb) [inline cEP-0025.md]: module should be first

File: cEP-0025.md [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+    """
+    BEAR_DEPS = {PyFlakesASTBear}
+
+    def remove_future_imports(self, file, lineno):
+        """
+        Removes all ImportFrom nodes from the input line
+        """
+        diff = Diff(file)
+        line = file[lineno - 1]
+        tree = ast.parse(line)
```
Comment (jayvdb) [inline cEP-0025.md]: Why parse needed. Use start and end line & column to remove?
Beware of `\` also

File: cEP-0025.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+                    origin=self,
+                    message=('Future import %s found' % node.name),
+                    file=filename,
+                    diffs={filename: corrected},
+                    line=result.get_node_location(node)[0])
+```
+
+## Demonstrating use of PyFlakesASTBear by creating a fixes bear
+
+A hasstle free solution to the problem of removing all the future imports
```
Comment (jayvdb) [inline cEP-0025.md]: `hassle`

**Review by jayvdb — CHANGES_REQUESTED** [review]
as already discussed, this needs a much simpler version of the fixes bear which doesnt use ast.parse.

Maybe it would be useful to show using ast.parse , but this isnt a valid scenario for it.

Also it would be useful to show what the the flake8 compatible plugin api would look like.  A example "NoFutureImportPlugin" would do perfectly to compare with the bear implementation.

File: cEP-0025.md [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+class NoFutureImport(object):
+    name = 'no_future'
+    version = __version__
+
+    def __init__(self, tree, filename):
+        self.tree = tree
+        self.filename = filename
+
+    def run(self):
+        pyflakes_result = Checker(self.tree, self.filename)
```
Comment (jayvdb) [inline cEP-0025.md]: We dont want every plugin to rebuild the pyflakes ast.

So we will need a new calling convention for these pyflakes plugins, and we will need to add some glue into flake8 to pass the Checker instance to the plugins.

A simple way to do this in a backwards conpatible way is for the plugin class to optionally have a 'set_pyflakes_data' which flake8 could call. And the plugin could build the pyflakes data itself if it wasnt given the data.

Then the plugin could work with coala, old flake8 and the "yet to be improved flake8"

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0025.md [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+class NoFutureImport(object):
+    name = 'no_future'
+    version = __version__
+
+    def __init__(self, tree, filename):
+        self.tree = tree
+        self.filename = filename
+        self.checker = None
+
+    def set_pyflakes_data(self, checker):
```
Comment (jayvdb) [inline cEP-0025.md]: change this to a `@property` called `checker`, and use `_checker` for storage

File: cEP-0025.md [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+        self.checker = checker
+
+    def run(self):
+        if self.checker is None:
+            self.set_pyflakes_data(Checker(self.tree, self.filename))
```
Comment (jayvdb) [inline cEP-0025.md]: as a `@property` , you can put this logic into the getter.

File: cEP-0025.md [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+        module_scope = list(filter(lambda scope: isinstance(scope,
+                                                            ModuleScope),
+                           pyflakes_result.deadScopes))[0]
+        for _, node in module_scope.items():
+            if isinstance(node, FutureImportation):
+                message = ('{code}: Future import {node} found'
+                           .format(node=node.name,
+                                   code=CODE))
+                yield node.source.lineno, node.source.col_offset, \
```
Comment (jayvdb) [inline cEP-0025.md]: please dont use `\`.

File: cEP-0025.md [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+
+    def run(self):
+        if self.checker is None:
+            self.set_pyflakes_data(Checker(self.tree, self.filename))
+        module_scope = list(filter(lambda scope: isinstance(scope,
+                                                            ModuleScope),
+                           pyflakes_result.deadScopes))[0]
+        for _, node in module_scope.items():
+            if isinstance(node, FutureImportation):
+                message = ('{code}: Future import {node} found'
```
Comment (jayvdb) [inline cEP-0025.md]: use `{name}` instead of `{node}`

File: cEP-0025.md [hunk 23]
```diff
@@ -0,0 +1,275 @@
...
+        self.filename = filename
+        self.checker = None
+
+    def set_pyflakes_data(self, checker):
+        self.checker = checker
+
+    def run(self):
+        if self.checker is None:
+            self.set_pyflakes_data(Checker(self.tree, self.filename))
+        module_scope = list(filter(lambda scope: isinstance(scope,
```
Comment (jayvdb) [inline cEP-0025.md]: you can also move this into a property getter to move the `run` look a lot simpler

File: cEP-0025.md (hunk 16 above)
Comment (jayvdb) [reply cEP-0025.md]: hmm.  three days ... no reply

Comment (ankitxjoshi) [reply cEP-0025.md]: Working on it. Cases like =>
```
from __future__ import print_function, \
                        unicode_literals 
```
are a real problem. Since the lineno for both the nodes are 1, 1 and not 1, 2 which I didn't expected.

File: cEP-0025.md (hunk 20 above)
Comment (MacBox7) [reply cEP-0025.md]: If we put this login in getter then won't it lead to infinite loop?

File: cEP-0025.md [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+
+    @property
+    def module_scope(self):
+        if self._module_scope is None:
+            self._module_scope = list(filter(lambda scope: isinstance(scope,
+                                                            ModuleScope),
+                                             self.checker.deadScopes))[0]
+        return self._module_scope
+
+    def set_pyflakes_data(self, checker):
```
Comment (jayvdb) [inline cEP-0025.md]: This is unnecessary now ?

Comment (ankitxjoshi) [reply cEP-0025.md]: Will flake8 use `setattr(foo, 'checker', instance)` to initialize the plugin with checker class? Or should we expose this function so that flake8 can use `set_pyflakes_data(instance)` to provide pyflakes AST?

Comment (jayvdb) [reply cEP-0025.md]: flake8 could use `foo.checker` , because it is a property

File: cEP-0025.md [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+
+## Introduction
+
+Flake8 is the most commonly used linter framework when it comes to static
+code analysis for python. It wraps around the tools pyflakes,
+pycodestyle and mccabe scripts. These tools are implemented on different
+interfaces and flake8 combines their potential.
+
+It even provides with a plugin mechanism. Additional linters are
+supported using the flake8 plugin system, and may be Python AST
```
Comment (jayvdb) [inline cEP-0025.md]: replace `AST` here , as it is actually the CPython `ast` standard library package which is used.
Make that very clear here, which helps show that the pyflakes AST is quite different.

File: cEP-0025.md [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+There has not been an option to use the pyflakes-enhanced-AST based checker
+API to create custom linters. Thus, the full potential of enhanced AST isn't
+utilized , a whole lot of rework is required to do the basic traversing
+and collection of important nodes. Pyflakes provides with a basic API
+that does the traversing. So, if a developer uses enhanced AST he just
+needs to work on the implementation of the new logic that his/her plugin
+provides and not about the fidelity of the basic node handlers.
+
+The reason for choosing pyflakes over its equivalents like
+[Asteroid](https://github.com/PyCQA/astroid) lies in its simplicity
```
Comment (jayvdb) [inline cEP-0025.md]: spelling

File: cEP-0025.md [hunk 27]
```diff
@@ -0,0 +1,306 @@
...
+and provision of lots of re-usable analysis. Asteroid is more detailed and
```
Comment (jayvdb) [inline cEP-0025.md]: spelling.

pls check all

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0025.md [hunk 28]
```diff
@@ -0,0 +1,306 @@
...
+Here is a brief overview of the architecture:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install pyflakes-bears`.
+2. The repository will contain a new package `pyflakes-bears` which
+   would house all bears which use pyflakes-enhanced-ast.
+3. The `pyflakes-bears` would contain a metabear `PyFlakesASTBear`
+   which would be used by all other plugin bears.
+4. The repository would even contain another package `pyflakes-generic-plugins`
+   that would be independent of coala and could be run using flake8.
```
Comment (jayvdb) [inline cEP-0025.md]: .. and could be run by flake8 or coala or any other similar tool.

File: cEP-0025.md [hunk 29]
```diff
@@ -0,0 +1,306 @@
...
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being
+used in `coala-bears` to test all its bears.
+
+## Implementation of PyFlakesASTBear
+
+1. PyFlakesASTBear will be LocalBear returning result of the type
```
Comment (jayvdb) [inline cEP-0025.md]: use backticks everywhere that a class name or literal string is mentioned

File: cEP-0025.md [hunk 30]
```diff
@@ -0,0 +1,306 @@
...
+   HiddenResult since these results are not meant for the users but
+   for the dependent bear. The result structure is described by
+   PyFlakesResult class.
+2. PyFlakesResult consists of a snapshot of the module at class,
+   function, module, generator ,and doctest level. Thus, providing
```
Comment (jayvdb) [inline cEP-0025.md]: misplaced `,`

File: cEP-0025.md [hunk 31]
```diff
@@ -0,0 +1,306 @@
...
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)
+        self.class_scopes = self.get_scopes(ClassScope, deadScopes)
+        self.function_scopes = self.get_scopes(FunctionScope, deadScopes)
+        self.generator_scopes = self.get_scopes(GeneratorScope, deadScopes)
+        self.doctest_scopes = self.get_scopes(DoctestScope, deadScopes)
```
Comment (jayvdb) [inline cEP-0025.md]: I suspect there may only be one doctest scope.
can you check this. if so, make it singular instead of plural, and list it under the module_scope variable

File: cEP-0025.md [hunk 32]
```diff
@@ -0,0 +1,306 @@
...
+```python
+# Imports not mentioned to maintain brevity
+
+class PyFlakesResult(HiddenResult):
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)
```
Comment (jayvdb) [inline cEP-0025.md]: I think this needs a `[0]` at the end to get only the first result.

File: cEP-0025.md [hunk 33]
```diff
@@ -0,0 +1,306 @@
...
+            dependency_results=dict()
+            ):
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            for node in result.get_nodes(result.module_scope[0],
+                                         FutureImportation):
+                lineno = result.get_node_location(node)[0]
+                corrected = self.remove_import(file, lineno, node.name)
+                yield Result.from_values(
+                    origin=self,
+                    message=('Future import %s found' % node.name),
```
Comment (jayvdb) [inline cEP-0025.md]: round brackets not needed.

File: cEP-0025.md [hunk 34]
```diff
@@ -0,0 +1,306 @@
...
+            ):
+        """
+        Uses PyFlakesASTBear to get only FutureImportation
+        """
+        # Preprocess all the lines containing __future__ imports
+        affected_lines = set()
+        # Get the maximum column offset of FutureImportation for a line
+        max_column_for_a_line = dict()
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            for node in result.get_nodes(result.module_scope[0],
```
Comment (jayvdb) [inline cEP-0025.md]: yup; `[0]` needed in the Result subclass, not here.

File: cEP-0025.md [hunk 35]
```diff
@@ -0,0 +1,306 @@
...
+                                         FutureImportation):
+                lineno = result.get_node_location(node)[0]
+                col_offset = result.get_node_location(node)[2]
```
Comment (jayvdb) [inline cEP-0025.md]: this `[2]` here isnt self-explanatory.

maybe get_node_location should return something which is easier to use.

File: cEP-0025.md [hunk 36]
```diff
@@ -0,0 +1,306 @@
...
+                affected_lines.add(lineno)
+
+        for line in affected_lines:
+            corrected = self.remove_future_imports(
+                        file, line, max_column_for_a_line[lineno],
+                        affected_lines
+                        )
+            yield Result.from_values(
+                origin=self,
+                message=('Future import/imports found'),
```
Comment (jayvdb) [inline cEP-0025.md]: Simplify to be 'Future import(s) found'

File: cEP-0025.md [hunk 37]
```diff
@@ -0,0 +1,306 @@
...
+                file=filename,
+                diffs={filename: corrected},
+                line=line)
+```
+
+## A generic implementation for detecting `__future__` imports
+
+This implementation doesn't requires coala to be installed and is
+designed so that it can be used in other tools like flake8. This
```
Comment (jayvdb) [inline cEP-0025.md]: clarify that coala will also be able to use these plugins, but these are also usable by flake8.

Also this example should be above the 'fixes' bear.

The fixes bear is the finale -- it shows why we would use coala bears -- there are features of coala which bears can use that generic plugins can not use.

File: cEP-0025.md (hunk 31 above)
Comment (ankitxjoshi) [reply cEP-0025.md]: No. There exists multiple `doctest_scope`. 1 scope per doctest. That's how it works.

File: cEP-0025.md (hunk 35 above)
Comment (ankitxjoshi) [reply cEP-0025.md]: We could even opt to remove the wrapper function.
`node.source.lineno` or `node.source.col__offset` will be good.

File: cEP-0025.md (hunk 31 above)
Comment (jayvdb) [reply cEP-0025.md]: Eww. ok

File: cEP-0025.md [hunk 38]
```diff
@@ -0,0 +1,305 @@
...
+        self.doctest_scopes = self.get_scopes(DoctestScope, deadScopes)
+        self.pyflakes_messages = pyflakes_messages
+
+    def get_scopes(self, scope_type, scopes):
+        return list(filter(lambda scope: isinstance(scope, scope_type),
+                           scopes))
+
+    def get_node_location(self, node):
+        return {'lineno': node.source.lineno,
+                'depth': node.source.depth - 1,
```
Comment (jayvdb) [inline cEP-0025.md]: `depth` isnt used in your examples.

and this method is just a thin wrapper around `node.source` ... maybe it should return `node.source`, which allows access to `lineno` and `col_offset`, and if pyflakes internals ever change then get_node_location` could return its own `NamedTuple` which provides those attributes.

But maybe we dont need `get_node_location`.  If the pyflakes internals ever removed or changed `node.source`, we could wrap all nodes with a backwards compatible class which provides a `node` and `node.source` which has the attributes which plugins expect.

File: cEP-0025.md [hunk 39]
```diff
@@ -0,0 +1,305 @@
...
+    """
+    BEAR_DEPS = {PyFlakesASTBear}
+
+    def run(self, filename, file,
+            dependency_results=dict()
+            ):
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            for node in result.get_nodes(result.module_scope,
+                                         FutureImportation):
+                lineno = result.get_node_location(node)['lineno']
```
Comment (jayvdb) [inline cEP-0025.md]: or this could be just `lineno = node.source.lineno`.  I am not seeing the benefit of `get_node_location`.

File: cEP-0025.md [hunk 40]
```diff
@@ -0,0 +1,305 @@
...
+                corrected = self.remove_import(file, lineno, node.name)
+                yield Result.from_values(
+                    origin=self,
+                    message='Future import %s found' % node.name,
+                    file=filename,
+                    diffs={filename: corrected},
+                    line=result.get_node_location(node)['lineno'])
```
Comment (jayvdb) [inline cEP-0025.md]: this could be `line=lineno`

File: cEP-0025.md [hunk 41]
```diff
@@ -0,0 +1,305 @@
...
+                                             self.checker.deadScopes))[0]
+        return self._module_scope
+
+    def run(self):
+        for _, node in self.module_scope.items():
+            if isinstance(node, FutureImportation):
+                message = '{code}: Future import {name} found'
+                           .format(name=node.name,
+                                   code=CODE)
+                yield (node.source.lineno, node.source.col_offset,
```
Comment (jayvdb) [inline cEP-0025.md:213]: I see you are using `node.source.foo` here, and it looks nice.

File: cEP-0025.md [hunk 42]
```diff
@@ -0,0 +1,305 @@
...
+        Uses PyFlakesASTBear to get only FutureImportation
+        """
+        # Preprocess all the lines containing __future__ imports
+        affected_lines = set()
+        # Get the maximum column offset of FutureImportation for a line
+        max_column_for_a_line = dict()
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            for node in result.get_nodes(result.module_scope,
+                                         FutureImportation):
+                lineno = result.get_node_location(node)['lineno']
```
Comment (jayvdb) [inline cEP-0025.md]: `get_node_location()` in this method also looks to be more complexity than is needed.

File: cEP-0025.md [hunk 43]
```diff
@@ -0,0 +1,305 @@
...
+                col_offset = result.get_node_location(node)['col_offset']
+                if lineno in max_column_for_a_line:
+                    max_column_for_a_line[lineno] = max(
+                                max_column_for_a_line[lineno],
```
Comment (jayvdb) [inline cEP-0025.md]: very strange indentation here.

File: cEP-0025.md [hunk 44]
```diff
@@ -0,0 +1,305 @@
...
+    def remove_future_imports(self, file, lineno,
+                              max_col_offset, affected_lines):
+        """
+        Removes all __future__ imports from the input line
+        """
+        diff = Diff(file)
+        line = file[lineno - 1]
+        # If a line contains \ in the end
+        if line.rstrip()[-1] == '\\':
+            next_line = file[lineno]
```
Comment (jayvdb) [inline cEP-0025.md]: and what happens if the `next_line` also contains `\` .. :P  (now you may understand why I hate `\` so passionately.

Some ways to handle this neatly is to create a concept of a 'logical line', which provides information about the entire line, and also have metadata about each distinct statement which gives `(start_line,start_col),(end_line,end_col)`.

Given the complexity of this, I suggest you add a code comment that the algorithm hasnt been adjusted to support `\` .  That will make the logic a lot simpler and then you can solve the `\` problem later, when you have real code and unit tests to verify all silly corner cases have been checked.

**Review by jayvdb — CHANGES_REQUESTED** [review]

**Review by jayvdb — APPROVED** [review]

**Comment by jayvdb** [comment]
ack https://github.com/coala/cEPs/pull/144/commits/95bc075d25b5603b3a0d5c2ab291c73f0e858897

**Comment by jayvdb** [comment]
@gitmate-bot ff
